haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,419 @@
+        ---
+         Otherwise, if the node is an output variable, store the result in the invocation context
+         Assign the value of the input variable from the invocation context to the output variable, store the
+            output variable in the invocation context, and pass it to the next pipeline node
+            ____
+            - name: shaper
+             params:
+               inputs:
+                   query:
+                     output: questions
+        ---
+
+        :param parent: The parent node of the current node.
+        :param node: A dictionary representing the function node.
+        :param invocation_context: A dictionary of variables available in the invocation context.
+        :raises ValueError: If the parameters specified in the YAML definition of the function are invalid.
+
+        """
+        if "func" in node:
+            self.invoke_function(parent, node, invocation_context)
+
+        elif "output" in node and parent in invocation_context:
+            invocation_context[node["output"]] = copy.deepcopy(invocation_context[parent])
+        else:
+            params_is_parent = "params" == parent
+            root_node = parent is None and all(key for key in node.keys() if key in invocation_context)
+            # if not a function, valid output, or params definition then this must be a root node in the inputs
+            # otherwise, something is wrong, raise an exception
+            if not params_is_parent and not root_node:
+                raise ValueError(
+                    f"Invalid input configuration. The input configuration should be a function invocation, or "
+                    f"an assignment. The input configuration is {node}"
+                )
+
+    def invoke_function(self, parent: str, node: Dict[str, Any], invocation_context: Dict[str, Any]) -> None:
+        """
+        Invoke a function on the given invocation context and stores the result back in the invocation context.
+
+        :param parent: The parent node of the current node.
+        :param node: A dictionary representing the function node.
+        :param invocation_context: A dictionary of variables available in the invocation context.
+        :raises ValueError: If the parameters specified in the YAML definition of the function are invalid.
+        """
+        params = node.get("params", [])
+        output = node.get("output", parent)
+        # use case when the function is invoked with positional arguments
+        if isinstance(params, list):
+            # resolve the argument values from the invocation context
+            args = [invocation_context[param] for param in params if param in invocation_context]
+
+            if len(args) != len(params):
+                raise ValueError(
+                    f"Invalid function arguments. The function {node['func']} "
+                    f"specifies in its YAML definition {len(params)} arguments yet {len(args)} arguments "
```

Comment:
Can the `Shaper` only be used with YAML? Or can it also be defined in code (I think so). If so, then we shouldn't reference YAML in the error message.